File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/java/exercises/practice/ledger/src/main/java/Ledger.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Comparator;

public class Ledger {
    public LedgerEntry createLedgerEntry(String d, String desc, int c) {
        LedgerEntry le = new LedgerEntry();
        le.setChange(c);
        le.setDescription(desc);
        le.setLocalDate(LocalDate.parse(d));
        return le;
    }

    public String format(String cur, String loc, LedgerEntry[] entries) {
        validateInputs(cur, loc);
        FormatConfig config = getFormatConfig(cur, loc);
        
        StringBuilder sb = new StringBuilder(config.header());

        if (entries.length > 0) {
            List<LedgerEntry> sortedEntries = getSortedEntries(entries);

            for (LedgerEntry e : sortedEntries) {
                String date = e.getLocalDate().format(DateTimeFormatter.ofPattern(config.datePattern()));

                String desc = e.getDescription();
                if (desc.length() > 25) {
                    desc = desc.substring(0, 22);
                    desc = desc + "...";
                }

                // Convert cents to currency units
                double change = e.getChange() / 100.0;
                String converted = String.format("%.02f", Math.abs(change));
                boolean isNegative = change < 0;

                String[] parts = converted.split("\\.");
                String amount = "";
                int count = 1;
                for (int ind = parts[0].length() - 1; ind >= 0; ind--) {
                    if (((count % 3) == 0) && ind > 0) {
                        amount = config.thousandSeparator() + parts[0].charAt(ind) + amount;
                    } else {
                        amount = parts[0].charAt(ind) + amount;
                    }
                    count++;
                }

                if (loc.equals("nl-NL")) {
                    amount = config.currencySymbol() + " " + amount + config.decimalSeparator() + parts[1];
                } else {
                    amount = config.currencySymbol() + amount + config.decimalSeparator() + parts[1];
                }
                
                if (isNegative && loc.equals("en-US")) {
                    amount = "(" + amount + ")";
                } else if (isNegative && loc.equals("nl-NL")) {
                    amount = "-" + amount + " ";
                } else if (loc.equals("nl-NL")) {
                    amount = " " + amount + " ";
                } else {
                    amount = amount + " ";
                }
                
                sb.append('\n');
                sb.append(String.format("%s | %-25s | %13s",
                        date,
                        desc,
                        amount));
            }

        }

        return sb.toString();
    }

    private record FormatConfig(
        String header,
        String datePattern,
        String currencySymbol,
        String decimalSeparator,
        String thousandSeparator
    ) {}
    
    private void validateInputs(String currency, String locale) {
        if (!currency.equals("USD") && !currency.equals("EUR")) {
            throw new IllegalArgumentException("Invalid currency");
        }
        if (!locale.equals("en-US") && !locale.equals("nl-NL")) {
            throw new IllegalArgumentException("Invalid locale");
        }
    }
    
    private FormatConfig getFormatConfig(String currency, String locale) {
        if (currency.equals("USD")) {
            return locale.equals("en-US") ?
                new FormatConfig(
                    "Date       | Description               | Change       ",
                    "MM/dd/yyyy", "$", ".", ","
                ) :
                new FormatConfig(
                    "Datum      | Omschrijving              | Verandering  ",
                    "dd/MM/yyyy", "$", ",", "."
                );
        }
        return locale.equals("en-US") ?
            new FormatConfig(
                "Date       | Description               | Change       ",
                "MM/dd/yyyy", "€", ".", ","
            ) :
            new FormatConfig(
                "Datum      | Omschrijving              | Verandering  ",
                "dd/MM/yyyy", "€", ",", "."
            );
    }
    
    private List<LedgerEntry> getSortedEntries(LedgerEntry[] entries) {
        List<LedgerEntry> negatives = new ArrayList<>();
        List<LedgerEntry> positives = new ArrayList<>();
        
        for (LedgerEntry entry : entries) {
            (entry.getChange() < 0 ? negatives : positives).add(entry);
        }
        
        negatives.sort(Comparator.comparing(LedgerEntry::getLocalDate));
        positives.sort(Comparator.comparing(LedgerEntry::getLocalDate));
        
        List<LedgerEntry> combined = new ArrayList<>(negatives);
        combined.addAll(positives);
        return combined;
    }

    public static class LedgerEntry {
        LocalDate localDate;
        String description;
        int change;

        public LocalDate getLocalDate() {
            return localDate;
        }

        public void setLocalDate(LocalDate localDate) {
            this.localDate = localDate;
        }

        public String getDescription() {
            return description;
        }

        public void setDescription(String description) {
            this.description = description;
        }

        public double getChange() {
            return change;
        }

        public void setChange(double change) {
            this.change = change;
        }
    }

}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/ledger/ledger.go
package ledger

import (
	"errors"
	"fmt"
	"sort"
	"strconv"
	"strings"
)

type Entry struct {
	Date        string // "Y-m-d"
	Description string
	Change      int // in cents
}

type EntriesByDateDesc []Entry

func getHeader(locale string) (string, error) {
	switch locale {
	case "nl-NL":
		return fmt.Sprintf("%-10s | %-25s | Verandering\n", "Datum", "Omschrijving"), nil
	case "en-US":
		return fmt.Sprintf("%-10s | %-25s | Change\n", "Date", "Description"), nil
	default:
		return "", errors.New("unsupported locale")
	}
}

func (e EntriesByDateDesc) Len() int { return len(e) }
func (e EntriesByDateDesc) Less(i, j int) bool {
	if e[i].Date != e[j].Date {
		return e[i].Date > e[j].Date
	}
	if e[i].Description != e[j].Description {
		return e[i].Description < e[j].Description
	}
	return e[i].Change < e[j].Change
}
func (e EntriesByDateDesc) Swap(i, j int) { e[i], e[j] = e[j], e[i] }

func parseDate(date string, locale string) (string, error) {
	if len(date) != 10 || date[4] != '-' || date[7] != '-' {
		return "", errors.New("invalid date format")
	}
	
	if locale == "nl-NL" {
		return date[8:10] + "-" + date[5:7] + "-" + date[:4], nil
	}
	return date[5:7] + "/" + date[8:10] + "/" + date[:4], nil
}

func truncateDescription(desc string) string {
	if len(desc) > 25 {
		return desc[:22] + "..."
	}
	return desc + strings.Repeat(" ", 25-len(desc))
}

func formatCurrency(cents int, currency string, locale string) (string, error) {
	// Implementation will be extracted next
	return "", nil // Temporary implementation to satisfy compiler
}

func FormatLedger(currency string, locale string, entries []Entry) (string, error) {
	var entriesCopy []Entry
	for _, e := range entries {
		entriesCopy = append(entriesCopy, e)
	}
	if len(entries) == 0 {
		if _, err := FormatLedger(currency, "en-US", []Entry{{Date: "2014-01-01", Description: "", Change: 0}}); err != nil {
			return "", err
		}
		return "", nil
	}
	// Sort entries by date, then description, then change
	impl := EntriesByDateDesc(entriesCopy)
	sort.Sort(impl)

	// Create header based on locale
	s, err := getHeader(locale)
	if err != nil {
		return "", err
	}
	// Parallelism, always a great idea
	co := make(chan struct {
		i int
		s string
		e error
	})
	for i, et := range entriesCopy {
		go func(i int, entry Entry) {
			if len(entry.Date) != 10 {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			d1, d2, d3, d4, d5 := entry.Date[0:4], entry.Date[4], entry.Date[5:7], entry.Date[7], entry.Date[8:10]
			if d2 != '-' {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			if d4 != '-' {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			de := entry.Description
			if len(de) > 25 {
				de = de[:22] + "..."
			} else {
				de = de + strings.Repeat(" ", 25-len(de))
			}
			var d string
			if locale == "nl-NL" {
				d = d5 + "-" + d3 + "-" + d1
			} else if locale == "en-US" {
				d = d3 + "/" + d5 + "/" + d1
			}
			negative := false
			cents := entry.Change
			if cents < 0 {
				cents = cents * -1
				negative = true
			}
			var a string
			if locale == "nl-NL" {
				if currency == "EUR" {
					a += "€"
				} else if currency == "USD" {
					a += "$"
				} else {
					co <- struct {
						i int
						s string
						e error
					}{e: errors.New("")}
				}
				a += " "
				centsStr := strconv.Itoa(cents)
				switch len(centsStr) {
				case 1:
					centsStr = "00" + centsStr
				case 2:
					centsStr = "0" + centsStr
				}
				rest := centsStr[:len(centsStr)-2]
				var parts []string
				for len(rest) > 3 {
					parts = append(parts, rest[len(rest)-3:])
					rest = rest[:len(rest)-3]
				}
				if len(rest) > 0 {
					parts = append(parts, rest)
				}
				for i := len(parts) - 1; i >= 0; i-- {
					a += parts[i] + "."
				}
				a = a[:len(a)-1]
				a += ","
				a += centsStr[len(centsStr)-2:]
				if negative {
					a += "-"
				} else {
					a += " "
				}
			} else if locale == "en-US" {
				if negative {
					a += "("
				}
				if currency == "EUR" {
					a += "€"
				} else if currency == "USD" {
					a += "$"
				} else {
					co <- struct {
						i int
						s string
						e error
					}{e: errors.New("")}
				}
				centsStr := strconv.Itoa(cents)
				switch len(centsStr) {
				case 1:
					centsStr = "00" + centsStr
				case 2:
					centsStr = "0" + centsStr
				}
				rest := centsStr[:len(centsStr)-2]
				var parts []string
				for len(rest) > 3 {
					parts = append(parts, rest[len(rest)-3:])
					rest = rest[:len(rest)-3]
				}
				if len(rest) > 0 {
					parts = append(parts, rest)
				}
				for i := len(parts) - 1; i >= 0; i-- {
					a += parts[i] + ","
				}
				a = a[:len(a)-1]
				a += "."
				a += centsStr[len(centsStr)-2:]
				if negative {
					a += ")"
				} else {
					a += " "
				}
			} else {
				co <- struct {
					i int
					s string
					e error
				}{e: errors.New("")}
			}
			var al int
			for range a {
				al++
			}
			co <- struct {
				i int
				s string
				e error
			}{i: i, s: d + strings.Repeat(" ", 10-len(d)) + " | " + de + " | " +
				strings.Repeat(" ", 13-al) + a + "\n"}
		}(i, et)
	}
	ss := make([]string, len(entriesCopy))
	for range entriesCopy {
		v := <-co
		if v.e != nil {
			return "", v.e
		}
		ss[v.i] = v.s
	}
	for i := 0; i < len(entriesCopy); i++ {
		s += ss[i]
	}
	return s, nil
}
